q2cli/util.py

```diff
@@ -234,3 +275,119 @@ def get_plugin_manager():
+    elif artifact.has_metadata():
+        try:
+            metadata = artifact.view(qiime2.Metadata)
+        except Exception as e:
+            header = ("There was an issue with viewing the artifact "
+                      "%s as QIIME 2 Metadata:" % fp)
+            tb = 'stderr' if '--verbose' in sys.argv else None
+            exit_with_error(e, header=header, traceback=tb)
+    else:
+        raise Exception("Artifacts with type %r cannot be viewed as"
+                        " QIIME 2 metadata:\n%r" % (artifact.type, fp))
+
+    return metadata
+
+
+def get_input(fp):
+    """ Gets a Result from a filepath if possible
+    """
+    import tempfile
+    import qiime2
+    import qiime2.sdk
+
+    get_plugin_manager()
+    try:
+        if ':' in fp:
+            artifact = convert_to_cache_input(fp)
+
+        # If we get here we either had a path without a ':' or we got
+        # None from convert_to_cache_input meaning the part of value
+        # before the ':' was not an existing cache
+        if ':' not in fp or artifact is None:
+            artifact = qiime2.sdk.Result.load(fp)
+    except OSError as e:
+        if e.errno == 28:
+            temp = tempfile.tempdir
+            raise ValueError(f'There was not enough space left on {temp!r} '
+                             f'to extract the artifact {fp!r}. (Try '
+                             'setting $TMPDIR to a directory with more '
+                             f'space, or increasing the size of {temp!r})')
+        else:
+            raise ControlFlowException(
+                '%r is not a QIIME 2 Artifact (.qza)' % fp) from e
+    except ValueError as e:
+        if 'does not contain the key' in str(e):
+            raise e
+        elif 'does not exist' in str(e):
+            # If value was also not an existing filepath
+            # containing a ':' we assume they wanted a cache
+            # but did not provide a valid one
+            if ':' in fp:
+                raise ValueError(f"The path {fp.split(':')[0]} is not a valid"
+                                 " cache") from e
+            else:
+                raise ValueError(f'{fp!r} is not a valid filepath') from e
+        else:
+            raise ControlFlowException(
+                '%r is not a QIIME 2 Artifact (.qza)' % fp) from e
+    # If we get here, all we really know is we failed to get a Result
+    except Exception as e:
+        raise ControlFlowException(
+            'There was a problem loading %s as a QIIME 2 Result: ' % fp) from e
+
+    return artifact
+
+
+def convert_to_cache_input(fp):
+    """ Determine if an input is in a cache and load it from the cache if it is
+    """
+    import os
+    from pathlib import Path
+    from qiime2.core.cache import Cache
+
+    split_path = fp.split(':')
+
+    # Handle the potential for : in the cache path
+    cache_path = ':'.join(elem for elem in split_path[:-1])
+    key = split_path[-1]
```

Comment:
I would create a helper using the aforementioned `rsplit` method to factor this shared bit out.